File: Algorithm_Swift/Algorithm_Swift/Sort/01_bubbleSort/BubbleSort.swift
//
//  BubbleSort.swift
//  Algorithm_Swift
//
//  Created by CHM on 2020/8/4.
//  Copyright © 2020 CHM. All rights reserved.
//

import Foundation

class BubbleSort {
    private static func bubbleSort<T>(_ array: inout Array<T>) where T: Comparable {
        guard !array.isEmpty else { return }
        
        let count = array.count
        var k = count - 1
        for _ in 0..<count-1 {
            var noExchange = true
            var minIndex = 0
            for j in 0..<k {
                if array[j] > array[j + 1] {
                    array.swapAt(j, j + 1)
                    noExchange = false // 记录本次循环是否发生过交换，如果没有，则表示当前数组已经是有序的了
                    minIndex = j // 用于记录本次循环发生最后一次交换时的最小下标
                }
            }
            if noExchange { break }
            k = minIndex
        }
    }
    
    private static func test<T>(_ testName: String, _ array: inout Array<T>) where T: Comparable {
        print("\(testName) begins:")
        print(array)
        bubbleSort(&array)
        print(array)
    }
    
    // 随机
    private static func test1() {
        var array = [4, 6, 1, 2, 9, 10, 20, 1, 1]
        test("test1", &array);
    }
    
    // 升序
    private static func test2() {
        var array = [1, 2, 3, 4, 5, 6, 7, 8]
        test("test2", &array);
    }
    
    // 降序
    private static func test3() {
        var array = [5, 4, 3, 2, 1]
        test("test3", &array);
    }
    
    // 有负数
    private static func test4() {
        var array = [-4, 6, -2, 9, 10, -20, 1, -1]
        test("test4", &array);
    }
    
    // 一个元素
    private static func test5() {
        var array = [4, 2]
        test("test5", &array);
    }
    
    static func Test() {
        test1()
        test2()
        test3()
        test4()
        test5()
    }
}

File: OfferReview/OfferReview/Sort/01_bubbleSort/BubbleSort.cpp
//
//  BubbleSort.cpp
//  OfferReview
//
//  Created by CHM on 2020/8/4.
//  Copyright © 2020 CHM. All rights reserved.
//

#include "BubbleSort.hpp"

int sum(int n) {
    return n > 1? sum(n - 1) + n: 1;
}

void BubbleSort::bubbleSort(int nums[], int count) {
    if (nums == nullptr || count <= 0) {
        return;
    }
    
    int k = count - 1;
    for (int i = 0; i < count - 1; ++i) {
        bool noExchange = true;
        int n = 0;
        for (int j = 0; j < k; ++j) {
            if (nums[j] > nums[j + 1]) {
                swap(&nums[j], &nums[j + 1]);
                noExchange = false;
                n = j;
            }
        }
        if (noExchange) {
            break;
        }
        k = n;
    }
    
    //if (nums == nullptr || count <= 0) {
    //    return;
    //}
    //
    //int k = count - 1; // 第二层循环的边界
    //for (int i = 0; i < count - 1; ++i) {
    //    bool noExchange = true; // 标记一趟比较操作下来是否发生过交换，如果没有的话表示当前序列已经有序
    //    int n = 0; // 记录一趟排序中最后一次交换的两个相邻元素的最小下标，此后该最小下标以后的数据就都是有序的
    //    for (int j = 0; j < k; ++j) {
    //        if (nums[j] > nums[j + 1]) {
    //            swap(&nums[j], &nums[j + 1]);
    //            noExchange = false;
    //            n = j;
    //        }
    //    }
    //
    //    if (noExchange) {
    //        break;
    //    }
    //    k = n; // 更新 k 值
    //}
}

// 测试代码
void BubbleSort::test(char* testName, int nums[], int count) {
    printf("%s begins: \n", testName);
    printArray("", nums, count);
    bubbleSort(nums, count);
    printArray("", nums, count);
}

void BubbleSort::test1() {
    int nums[9] = {4, 6, 1, 2, 9, 10, 20, 1, 1};
    test("test1", nums, 9);
}

void BubbleSort::test2() {
    int nums[8] = {1, 2, 3, 4, 5, 6, 7, 8};
    test("test2", nums, 8);
}

void BubbleSort::test3() {
    int nums[5] = {5, 4, 3, 2, 1};
    test("test3", nums, 5);
}

void BubbleSort::test4() {
    int nums[8] = {-4, 6, -2, 9, 10, -20, 1, -1};
    test("test4", nums, 8);
}

void BubbleSort::test5() {
    int nums[2] = {4, 2};
    test("test5", nums, 2);
}

void BubbleSort::test6() {
    int nums[] = {};
    test("test5", nums, 0);
}


void BubbleSort::Test() {
    test1();
    test2();
    test3();
    test4();
    test5();
    test6();
}

//if (nums == nullptr || count <= 0) {
//    return;
//}
//
//int k = count - 1; // 第二层循环的边界
//for (int i = 0; i < count - 1; ++i) {
//    bool noExchange = true; // 标记一趟比较操作下来是否发生过交换，如果没有的话表示当前序列已经有序
//    int n = 0; // 记录一趟排序中最后一次交换的两个相邻元素的最小下标，此后该最小下标以后的数据就都是有序的
//    for (int j = 0; j < k; ++j) {
//        if (nums[j] > nums[j + 1]) {
//            swap(&nums[j], &nums[j + 1]);
//            noExchange = false;
//            n = j;
//        }
//    }
//
//    if (noExchange) {
//        break;
//    }
//    k = n; // 更新 k 值
//}
